Code/GraphMol/Descriptors/AtomicEnvironmentVector.cpp

```diff
@@ -0,0 +1,679 @@
+  double zeta = 32;
+  double etaA = 8;
+
+  auto distances12 = vectors12.matrix().rowwise().norm().array();
+
+  ArrayXXd cosineAngles(vectors12.rows() / 2, 1);
+  for (auto i = 0; i < vectors12.rows() / 2; i++) {
+    cosineAngles(i, 0) = vectors12.matrix().row(i).dot(
+        vectors12.matrix().row(i + vectors12.rows() / 2));
+    cosineAngles(i, 0) =
+        0.95 * cosineAngles(i, 0) /
+        (vectors12.matrix().row(i).norm() *
+         vectors12.matrix().row(i + vectors12.rows() / 2).norm());
+  }
+  auto angles = cosineAngles.acos();
+  auto fcj12 = CosineCutoff(distances12, cutoff);
+
+  ArrayXXd factor1(angles.rows(), 8);
+  for (auto i = 0; i < angles.rows(); i++) {
+    factor1.row(i) =
+        (((-1 * (ShfZ.rowwise() - angles.row(i)).array()).cos().array() + 1)
+             .array() /
+         2)
+            .array()
+            .pow(zeta);
+  }
+
+  ArrayXXd distance12sum(distances12.rows() / 2, 1);
+
+  for (auto i = 0; i < distances12.rows() / 2; i++) {
+    distance12sum(i, 0) =
+        distances12(i, 0) + distances12(i + distances12.rows() / 2, 0);
+  }
+
+  ArrayXXd factor2(distance12sum.rows(), 4);
+  for (auto i = 0; i < distance12sum.rows(); i++) {
+    factor2.row(i) =
+        ((ShfA.rowwise() - (distance12sum.array() / 2).row(i)).pow(2).array() *
+         -etaA)
+            .exp();
+  }
+
+  ArrayXXd fcj12prod(fcj12.rows() / 2, 1);
+  for (auto i = 0; i < fcj12.rows() / 2; i++) {
+    fcj12prod(i, 0) = fcj12(i, 0) * fcj12(i + fcj12.rows() / 2, 0);
+  }
+
+  ArrayXXd ret(fcj12prod.rows(), 32);
+  for (auto i = 0; i < ret.rows(); i++) {
+    int idx = 0;
+    for (int j = 0; j < 8; j++) {
+      for (int k = 0; k < 4; k++) {
+        ret(i, idx) = 2 * factor1(i, j) * factor2(i, k) * fcj12prod(i, 0);
+        idx++;
+      }
+    }
+  }
+  return ret;
+}
+
+std::vector<int> cumsumFromZero(std::vector<int> count) {
+  std::vector<int> cumsum{0};
+  int index = 1;
+  for (size_t i = 0; i < count.size() - 1; i++) {
+    cumsum.push_back(cumsum[index - 1] + count[i]);
+    index++;
+  }
+
+  return cumsum;
+}
+
+//! Calculates triplets of atoms according to pairs of atoms close to each other
+/*!
+  \param atomIndex12Angular   Pairs of atoms close to each other according to
+  defined cutoff
+
+  \return pair of a vector containing indices of centrals atoms and Matrix
+  containing pairs of their neighbours
+
+  \verbatim
+    Input: indices for pairs of atoms that are close to each other.
+    each pair only appear once, i.e. only one of the pairs (1, 2) and
+    (2, 1) exists.
+
+    Output: indices for all central atoms and it pairs of neighbors. For
+    example, if input has pair (0, 1), (0, 2), (0, 3), (0, 4), (1, 2),
+    (1, 3), (1, 4), (2, 3), (2, 4), (3, 4), then the output would have
+    central atom 0, 1, 2, 3, 4 and for cental atom 0, its pairs of neighbors
+    are (1, 2), (1, 3), (1, 4), (2, 3), (2, 4), (3, 4)
+  \endverbatim
+*/
+std::pair<std::vector<int>, ArrayXXd> TripleByMolecules(
```

Comment:
This function includes a lot of code and very few inline comments explaining what's going on (even briefly).
That not only makes it hard to figure out what's going on while reviewing, but it will make maintenance/bug fixing hard as well.